lib/authz/permissions.go

```diff
@@ -289,6 +295,40 @@ func (a *authorizer) fromUser(ctx context.Context, userI interface{}) (*Context,
 	}
 }
 
+var ErrIPPinningMissing = trace.AccessDenied("pinned IP is required for the user, but is not present on identity")
+var ErrIPPinningMismatch = trace.AccessDenied("pinned IP doesn't match observed client IP")
+
+// CheckIPPinning verifies IP pinning for the identity, using the client IP taken from context.
+// Check is considered successful if no error is returned.
+func CheckIPPinning(ctx context.Context, identity tlsca.Identity, pinSourceIP bool) error {
+	if identity.PinnedIP == "" {
+		if pinSourceIP {
+			return ErrIPPinningMissing
+		}
+		return nil
+	}
+
+	clientSrcAddr, err := ClientAddrFromContext(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clientIP, _, err := net.SplitHostPort(clientSrcAddr.String())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if clientIP != identity.PinnedIP {
+		logrus.WithFields(logrus.Fields{
```

Comment:
Hm, I think this is perfect place for this debug log, there's one direct caller - authorizer and myriad of higher level callers that don't do logging. And they actually don't really have direct access to the ip fields which we log here.

Reply:
This is a problem because you are using the default logger which is not configured with all the configurations provided in the config.

I still think the caller must handle the log here since it has access to the configured logger.

You can `return trace.Wrap(ErrIPPinningMismatch, "pinned IP and client IP mismatch: client_ip: %v, pinned_ip: %v", clientIP,identity.PinnedIP)`

With this, you still have errors.Is working and you can log it with the correct logger.

Reply:
I'm hesitant to return specific IPs in error for security concerns. But I agree that using default logger is not ideal. In this case main problem I think that we're missing component field. I've changed `CheckIPPinning` to accept logger and currently authorizer will just provide logger with `trace.Component: "authorizer"` and I will handle adding logger to `NewAuthorizer` in a separate PR (issue for this https://github.com/gravitational/teleport/issues/25035 )